test_conformance/extensions/cl_khr_command_buffer/command_buffer_printf.cpp

```diff
@@ -0,0 +1,556 @@
+            R"(
+      __kernel void print(__global char* in, __global char* out, __global int* offset)
+      {
+          size_t id = get_global_id(0);
+          int ind = offset[0] + offset[1] * id;
+          for(int i=0; i<offset[1]; i++) out[ind+i] = in[i];
+          printf("%s", in);
+      })";
+
+        error = create_single_kernel_helper_create_program(context, &program, 1,
+                                                           &kernel_str);
+        test_error(error, "Failed to create program with source");
+
+        error = clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+        test_error(error, "Failed to build program");
+
+        kernel = clCreateKernel(program, "print", &error);
+        test_error(error, "Failed to create print kernel");
+
+        return CL_SUCCESS;
+    }
+
+    //--------------------------------------------------------------------------
+    size_t data_size() const override
+    {
+        return sizeof(cl_char) * num_elements * buffer_size_multiplier
+            * max_pattern_length;
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int SetUpKernelArgs() override
+    {
+        cl_int error = CL_SUCCESS;
+
+        in_mem = clCreateBuffer(context, CL_MEM_READ_ONLY,
+                                sizeof(cl_char) * (max_pattern_length + 1),
+                                nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        out_mem = clCreateBuffer(context, CL_MEM_WRITE_ONLY, data_size(),
+                                 nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        cl_int offset[] = { 0, max_pattern_length };
+        off_mem =
+            clCreateBuffer(context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR,
+                           sizeof(offset), offset, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        error = clSetKernelArg(kernel, 0, sizeof(in_mem), &in_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        error = clSetKernelArg(kernel, 1, sizeof(out_mem), &out_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        error = clSetKernelArg(kernel, 2, sizeof(off_mem), &off_mem);
+        test_error(error, "clSetKernelArg failed");
+
+        return CL_SUCCESS;
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int SetUp(int elements) override
+    {
+        // Query if device supports simultaneous use
+        cl_device_command_buffer_capabilities_khr capabilities;
+        cl_int error =
+            clGetDeviceInfo(device, CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR,
+                            sizeof(capabilities), &capabilities, NULL);
+        test_error(
+            error,
+            "Unable to query CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR");
+
+        if ((capabilities & CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR)
```

Comment:
This is checking `CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR` rather than `CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR`.

If `CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR` isn't supported the printf tests should skip rather than fail.

Reply:
So although this check has now been updated to `CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR` I still don't think it's quite right. The test should skip rather than fail if this capability isn't supported by a device, so think this check should be in `CommandBufferPrintfTest::Skip()`

Reply:
hmm, then maybe I should add new flag "printf_use_support" and move this piece of code to parent structure, right ?

Reply:
Flag added, condition moved to `CommandBufferPrintfTest::Skip()` method.